packages/flutter/lib/src/widgets/navigator_pop_handler.dart

```diff
@@ -0,0 +1,111 @@
+///
+/// ** See code in examples/api/lib/widgets/navigator_pop_handler/navigator_pop_handler.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [PopScope], which allows toggling the ability of a [Navigator] to
+///    handle pops.
+///  * [NavigationNotification], which indicates whether a [Navigator] in a
+///    subtree can handle pops.
+class NavigatorPopHandler extends StatefulWidget {
+  /// Creates an instance of [NavigatorPopHandler].
+  const NavigatorPopHandler({
+    super.key,
+    this.onPop,
+    this.enabled = true,
+    required this.child,
+  });
+
+  /// The widget to place below this in the widget tree.
+  ///
+  /// Typically this is a [Navigator] that will handle the pop when [onPop] is
+  /// called.
+  final Widget child;
+
+  /// Whether this widget's ability to handle system back gestures is enabled or
+  /// disabled.
+  ///
+  /// When false, there will be no effect on system back gestures. If provided,
+  /// [onPop] will still be called.
+  ///
+  /// This can be used, for example, when the nested [Navigator] is no longer
+  /// active but remains in the widget tree, such as in an inactive tab.
+  ///
+  /// Defaults to true.
+  final bool enabled;
+
+  /// Called when a handleable pop event happens.
+  ///
+  /// For example, a pop is handleable when a [Navigator] in [child] has
+  /// multiple routes on its stack. It's not handleable when it has only a
+  /// single route, and so [onPop] will not be called.
+  ///
+  /// Typically this is used to pop the [Navigator] in [child]. See the sample
+  /// code on [NavigatorPopHandler] for a full example of this.
+  final VoidCallback? onPop;
+
+  @override
+  State<NavigatorPopHandler> createState() => _NavigatorPopHandlerState();
+}
+
+class _NavigatorPopHandlerState extends State<NavigatorPopHandler> {
+  bool _canPop = true;
+
+  @override
+  Widget build(BuildContext context) {
+    // When the widget subtree indicates it can handle a pop, disable popping
+    // here, so that it can be manually handled in canPop.
+    return PopScope(
+      canPop: !widget.enabled || _canPop,
+      onPopInvoked: (bool didPop) {
+        if (didPop) {
+          return;
+        }
+        widget.onPop?.call();
+      },
+      // Listen to changes in the navigation stack in the widget subtree.
+      child: NotificationListener<NavigationNotification>(
+        onNotification: (NavigationNotification notification) {
+          // If this subtree cannot handle pop, then set canPop to true so
+          // that our PopScope will allow the Navigator higher in the tree to
+          // handle the pop instead.
+          final bool nextCanPop = !notification.canHandlePop;
+          if (nextCanPop != _canPop) {
+            // It's possible to receive a NavigationNotification during a build,
+            // so wait until the next frame to call setState.
+            SchedulerBinding.instance.addPostFrameCallback((Duration timeStamp) {
```

Comment:
alternatively the Notification can be sent as postframecallback in Navigator since it will have to be handle in postframecallback anyway (except for the root one that can be handle regardless the phase)

Reply:
I'm going to check the scheduler phase here in NavigatorPopHandler and keep the addPostFrameCallback here.  I just tried out doing it in Navigator but the problem is that the Navigator may navigate and then unmount, so that when the callback is executed, the Navigator is unmounted and I can't dispatch the notification.  However, other widgets may care about that last notification before unmount.  So I think it should be up to the receiver to check the phase if necessary.

Sending multiple notifications in a single frame should be handled correctly now because, even when they are handled in post frame callbacks, they will be handled in order (I tried it to make sure).  We could deduplicate/debounce them as a small improvement but it's probably not worth it, and some people might want to receive all the notifications anyway.

Reply:
shouldn't the unmount itself send the latest update? If Navigator use postframecallback, it could also check for mounted before sending out the pending update?

Either way I am fine with handling in NavigatorPopHandler

Reply:
I was able to get it to work with the postFrameCallback in Navigator with all of my examples 👍.